Reserved Python keywords support (Issue 105)

javiersorianoruiz opened the pull request:
We (@Bit-Seq, @jordialexalar and me) have implemented a solution for the issue 105.

https://github.com/eclipse-cyclonedds/cyclonedds-python/issues/105

With this change, now, the IDL file can contain reserved Python keywords and when this IDL file is translated into Python the IDLC Python plugin has been modified to add a prefix "_" to the found reserved keywords.

We have had into account all the posible declaration using IDL: structs, modules, enums, typedefs...

We have tested broadly this change. The developed and executed tests can be review in detail in this git repository:

https://github.com/javiersorianoruiz/cyclonedds-python-checkReservedKeyword

These are not the formal tests that maybe are required by cyclonedds-python project, but they have allowed us to verify that the implementation is right and have no colateral issues.

Maybe, It would be needed to develop another test batery based on pytest and integrated into cyclonedds-python project, following your testing architecteture, What do you think?, Is it needed?

thijsmie commented on cyclonedds/idl/_xt_builder.py:
```diff
@@ -974,9 +974,19 @@ def _xt_minimal_struct_member(cls, entity: Type[IdlStruct], name: str, _type: ty
     @classmethod
     def _xt_complete_struct_member(cls, entity: Type[IdlStruct], name: str, _type: type) -> xt.CompleteStructMember:
+        #Reserved Python keywords support (Issue 105)
+        realName=name
```
This needs a little bit of formatting and removing the duplicate dict access. I suggest the following:

````python
# Reserved Python keywords support (#105)
annotations = get_idl_field_annotations(entity).get(name)
real_name = annotations["name"] if annotations and "name" in annotations else name

javiersorianoruiz replied:
Hello @thijsmie , thank you for your fast answer.

About "we can also resolve types dynamically from the network using the XTInterpreter", How do we have to do this? How do we have to write the IDL file for the method XTInterpreter._make_complete_struct or XTInterpreter._make_complete_union to be executed during the translation from IDL to Python ?

If you give us more information about how to reproduce this conversion, we can try to implement the change ourselves as we have done with the rest of cases.


About your suggestion "annotations = get_idl_field_annotations(entity).get(name) ...", you are right, your implementation is smarter, we will change it.

thijsmie commented on idlpy/src/context.c:
```diff
@@ -530,7 +534,16 @@ static idl_retcode_t write_module_headers(FILE *fh, idlpy_ctx octx, idlpy_module
-            idl_fprintf(fh, fmt_entity, i > 0 ? ", " : "", idlpy_ssos_at(ctx->this_idl_file->entities, i));
+
+           //Reserved Python keywords support (Issue 105)
+            const char *name = idlpy_ssos_at(ctx->this_idl_file->entities, i);
```
If you change `get_c_name` to return the name unmodified when it is not a keyword you can write:
```c
const char *name = get_c_name(idlpy_ssos_at(ctx->this_idl_file->entities, i));
```
However, I suspect the `ctx->this_idl_file->entities` SSOS should contain already filtered names, so here the call is unnecessary?

javiersorianoruiz replied:
OK, we can do as you suggest, but the cyclonedds-cxx implementation (regarding also using C++ reserved keywords) implements this function (file `idlpy/src/types.c`):

```
const char *get_c_name(const char *name)
{
  //printf("get_c_name: %s \n", name);
  static const size_t n = sizeof(python_keywords)/sizeof(python_keywords[0]);

  /* search through the Python keyword list */
  for (size_t i=0; i < n; i++) {
    if (strcmp(python_keywords[i] + 1, name) == 0)
      return python_keywords[i];
  }

  return NULL;
}
```

**In order to be consistent with other implementations (cxx versus Python), we have followed this example, so, what do you think we have to do at the end?**


Regarding to "_However, I suspect the ctx->this_idl_file->entities SSOS should contain already filtered names, so here the call is unnecessary?_"  

Yes indeed,  `ctx->this_idl_file->entities` does contain the filtered names, so here, the call is unnecessary, in one of our tests, we have worked with this IDL file:

```
module pass {
    struct struct_test {
        char var;
    };
};
```

Our version of the plugin Python IDLC compiler, generates a folder con name "_pass" and inside the files:

```
__init__.py

from ._test import struct_test
__all__ = ["struct_test", ]
```
```
_test.py
from enum import auto
from typing import TYPE_CHECKING, Optional
from dataclasses import dataclass

import cyclonedds.idl as idl
import cyclonedds.idl.annotations as annotate
import cyclonedds.idl.types as types

# root module import for resolving types
import _pass


@dataclass
@annotate.final
@annotate.autoid("sequential")
class struct_test(idl.IdlStruct, typename="_pass.struct_test"):
    var: types.char
```

So, our version is working properly, and we have tested it with a publisher and subscriber in order to check that a message with this format can be sent and received.

thijsmie replied:
There is indeed no need to follow the C++ code here. We're going for maximum readability always, if it is used differently in one other place that doesn't really matter to me.

thijsmie commented on idlpy/src/types.c:
```diff
@@ -25,7 +25,32 @@
+#undef _py_
+};
+
+const char *get_c_name(const char *name)
```
I think `get_c_name` is not very descriptive to what it actually does: escape python keywords. I think it is actually prudent to define `idlpy_identifier` that calls `idl_identifier` and then filters keywords. This simplifies your code in quite a few places since there is no modified vs non-modified handling needed.

There are a couple places where a name of something that is not a node is determined. For that you can define a `filter_python_keywords` function and use that inside `idlpy_identifier`.

There is a few non-keyword things you should also add to the filter, namely `idl`, `annotate`, `types`, `auto`, `TYPE_CHECKING`, `Optional` and `dataclass` since they are occuring as imports at the top of any idl python output. See also https://github.com/eclipse-cyclonedds/cyclonedds-python/issues/161.

javiersorianoruiz replied:
Ok, so, we understand you don't want us to follow the cxx way of implementation. 

To confirm with you:

1. Modify name of function `get_c_name `to `filter_python_keywords `
2. `filter_python_keywords` never returns NULL, returns the original name or the original name with the prefix "_".
3. We implement a new function `const char *idlpy_identifier(const void *node)` which internally is going to call:

```
   name = idl_identifier(node) 
   name = filter_python_keywords(name)
   return  name;
```

About more keywords, OK, we will also review this #161 to add all the required keywords. We have filtered the Python keywords indicated in this web page:

https://www.w3schools.com/python/python_ref_keywords.asp
https://www.geeksforgeeks.org/python-keywords/

But, only for curiosity, why do we have to add other keywords that are not considered reserved?, We suppose that the Python code generated doesn't work if it contains any of the "extra" keywords, does it?

thijsmie replied:
The proposed modifications sound okay to me.

As for your questions, indeed, they are not keywords but are used in the generated code:

```python
from enum import auto
from typing import TYPE_CHECKING, Optional
from dataclasses import dataclass

import cyclonedds.idl as idl
import cyclonedds.idl.annotations as annotate
import cyclonedds.idl.types as types
```

thijsmie commented on idlpy/src/types.c:
```diff
@@ -146,13 +179,21 @@ emit_field(
+
     if (idl_is_default_case(parent)) {
+
+        //printf("emit_field idl_is_default_case name : %s , type %s \n", name , type);
```
Please don't leave commented code in the PR.

javiersorianoruiz replied:
ha ha ha, it was an oversight, we have already deleted this printf and another one that we have also found

thijsmie commented on idlpy/src/types.c:
```diff
@@ -146,13 +179,21 @@ emit_field(
     }
     else if (idl_is_case(parent)) {
+
+        //printf("emit_field idl_is_case name : %s , type %s \n", name , type);
```
Please don't leave commented code in the PR.

javiersorianoruiz replied:
all the printfs that we have added have been deleted

thijsmie commented on idlpy/src/types.c:
```diff
@@ -175,6 +216,9 @@ emit_field(
 
     if (idl_is_member(parent)) {
+
+        //printf("emit_field idl_is_member name : %s , type %s \n", name , type);
```
Please don't leave commented code in the PR.

javiersorianoruiz replied:
all the printfs that we have added have been deleted

thijsmie commented on idlpy/src/types.c:
```diff
@@ -185,42 +229,68 @@ emit_field(
+    
+    idlpy_ctx_printf(ctx, "\n    %s: %s", nameField, type);
+
+    if (nameFieldModified) {
```
If you switch everything to `idlpy_identifier` this will become `if (name != idl_identifier(node))`. Note that `strcmp` is not needed, we're comparing constant pointers and that is fine in this case.

thijsmie commented on idlpy/src/types.c:
```diff
@@ -185,42 +229,68 @@ emit_field(
 
     if (idl_is_member(parent)) {
+
+        //printf("emit_field idl_is_member2 name : %s , type %s \n", name , type);
```
Please don't leave commented code in the PR.

Bit-Seq replied:
After reviewing it today, all the printfs we have added have been deleted.

thijsmie commented on idlpy/src/types.c:
```diff
@@ -699,8 +848,19 @@ emit_const(
 
     const idl_literal_t *literal = ((const idl_const_t *)node)->const_expr;
 
+    //printf("------- emit_const: name %s \n",idl_identifier(node));
```
Please don't leave commented code in the PR.

Bit-Seq replied:
After reviewing it today, all the printfs we have added have been deleted.

thijsmie commented on idlpy/src/types.h:
```diff
@@ -17,4 +17,8 @@
 idl_retcode_t generate_types(const idl_pstate_t *pstate, idlpy_ctx ctx);
 
+//Reserved Python keywords support (Issue 105)
+const char *get_c_name(const char *name);
```
Maybe this function should live in `naming.h` instead?

Bit-Seq replied:
Ok. That header seems more proper to contain the method. We will move it.

thijsmie reviewed (changes requested):
Hi @javiersorianoruiz,

This looks already really nice! I have a couple things that need some attention, mostly stylistic. There is one feature you missed, we can also resolve types dynamically from the network using the `XTInterpreter`. The names in `XTInterpreter._make_complete_struct`, `XTInterpreter._make_complete_union` etc also need to take care with python keywords. I can also handle this myself in a followup PR, just let me know.

Bit-Seq commented:
Another detail not mentioned above. True and False keywords are not supported, due to they are also keywords in the IDL.

The IDLC compiler returns a syntax error even before trying to convert the IDL file to Python, so IDLC does not support these words in an IDL file for any supported language.

Anyway, True and False are keywords in almost all languages, and it is rare a lot to try to define an object with those identifiers.

javiersorianoruiz commented:
We have updated the code with your suggested changes. So you can advance in the pull request and merge process.

Everything is working properly in our testing: https://github.com/javiersorianoruiz/cyclonedds-python-checkReservedKeyword

We have pending:

```
About "we can also resolve types dynamically from the network using the XTInterpreter", How do we have to do this? How do we have to write the IDL file for the method XTInterpreter._make_complete_struct or XTInterpreter._make_complete_union to be executed during the translation from IDL to Python?

If you give us more information about how to reproduce this conversion, we can try to implement the change ourselves as we have done with the rest of cases.
``` 

We have added also the keywords you ask for them, so, consider closing Issue #161  if the pull request is ok for you.

thijsmie reviewed (approved):
I think it is okay like this! Thank you so much for the contribution. The last remaining point I mentioned about the network types is something I'll tackle later.
